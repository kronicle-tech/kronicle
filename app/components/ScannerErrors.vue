<template>
  <!-- eslint-disable vue/require-v-for-key -->
  <ul v-if="formattedScannerErrors && formattedScannerErrors.length > 0">
    <li v-for="scannerError in formattedScannerErrors">{{ scannerError }}</li>
  </ul>
  <!-- eslint-enable -->
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { ScannerError } from '~/types/kronicle-service'

function formatScannerError(value: ScannerError) {
  let text = value.message
  if (value.cause) {
    text += ' | ' + formatScannerError(value.cause)
  }
  return text
}

export default Vue.extend({
  props: {
    scannerErrors: {
      type: Array as PropType<ScannerError[]>,
      default: undefined,
    },
  },
  computed: {
    formattedScannerErrors() {
      if (!this.scannerErrors) {
        return null
      }

      return this.scannerErrors.map(formatScannerError)
    },
  },
})
</script>
