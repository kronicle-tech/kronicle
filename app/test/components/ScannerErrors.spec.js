import { mount } from '@vue/test-utils'
import ScannerErrors from '@/components/ScannerErrors.vue'

describe('ScannerErrors', () => {
  let propsData
  let wrapper
  const createWrapper = () => {
    wrapper = mount(ScannerErrors, { propsData })
  }

  beforeEach(() => {
    propsData = {}
  })

  afterEach(() => {
    wrapper.destroy()
    wrapper = null
  })

  describe('when ScannerErrors prop is not set', () => {
    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when ScannerErrors prop is set to an empty array', () => {
    beforeEach(() => {
      propsData.scannerErrors = []
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when ScannerErrors prop is set to one scanner error', () => {
    beforeEach(() => {
      propsData.scannerErrors = [
        {
          message: 'Test Message 1',
        },
      ]
    })

    test('renders an unordered list showing the scanner error', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })

    describe('when the scanner error has a cause', () => {
      beforeEach(() => {
        propsData.scannerErrors[0].cause = {
          message: 'Test Cause Message 1',
        }
      })

      test('renders the scanner error including the cause', () => {
        createWrapper()
        expect(wrapper.html()).toMatchSnapshot()
      })

      describe('when the cause has its own cause', () => {
        beforeEach(() => {
          propsData.scannerErrors[0].cause.cause = {
            message: 'Test Cause of Cause Message 1',
          }
        })

        test('renders the scanner error including both causes', () => {
          createWrapper()
          expect(wrapper.html()).toMatchSnapshot()
        })
      })
    })
  })

  describe('when ScannerErrors prop is set to multiple ScannerErrors', () => {
    beforeEach(() => {
      propsData.scannerErrors = [
        {
          message: 'Test Message 1',
        },
        {
          message: 'Test Message 2',
        },
        {
          message: 'Test Message 3',
        },
      ]
    })

    test('renders an unordered list showing the ScannerErrors', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
