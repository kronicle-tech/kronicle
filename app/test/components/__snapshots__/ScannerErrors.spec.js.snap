// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScannerErrors when ScannerErrors prop is set to multiple ScannerErrors renders an unordered list showing the ScannerErrors 1`] = `
"<ul>
  <li>Test Message 1</li>
  <li>Test Message 2</li>
  <li>Test Message 3</li>
</ul>"
`;

exports[`ScannerErrors when ScannerErrors prop is set to one scanner error renders an unordered list showing the scanner error 1`] = `
"<ul>
  <li>Test Message 1</li>
</ul>"
`;

exports[`ScannerErrors when ScannerErrors prop is set to one scanner error when the scanner error has a cause renders the scanner error including the cause 1`] = `
"<ul>
  <li>Test Message 1 | Test Cause Message 1</li>
</ul>"
`;

exports[`ScannerErrors when ScannerErrors prop is set to one scanner error when the scanner error has a cause when the cause has its own cause renders the scanner error including both causes 1`] = `
"<ul>
  <li>Test Message 1 | Test Cause Message 1 | Test Cause of Cause Message 1</li>
</ul>"
`;
